### PR TITLE
Remove caseSensitivityKey invalidation

### PR DESCRIPTION
**Pull request by akshya96**
Jira issue: https://hashicorp.atlassian.net/browse/VAULT-16845
Ent PR: https://github.com/hashicorp/vault-enterprise/pull/4233

A problem occurs when the caseSensitivityKey is invalidated due to a change in its value from True to False in storage, and duplicate groups are discovered. As a result, the loadGroups function returns an error, preventing the loading of groups and potentially causing missing groups in the memDB.

To resolve this issue, the proposed solution is to eliminate the invalidation case for the caseSensitivityKey. Since the value in storage can only be changed on the primary active node during initialization, there is no need for further invalidation. A warning is anyways being logged in loadIdentityStoreArtifacts function during the unseal process before initialization, if duplicate groups are encountered.

### CODE REVIEW

**Comment by davidadeleon** [comment]
If we are removing the invalidation for this key in storage, should we also look to remove the key from storage in general moving forward? The only place we update this value in storage is here: https://github.com/hashicorp/vault/blob/0d147185d6304838d4667ef7cb30d8ed3a41da3f/vault/identity_store.go#L649-L656
Which looks to also be (after this change) the last reference for this struct: https://github.com/hashicorp/vault/blob/0d147185d6304838d4667ef7cb30d8ed3a41da3f/vault/identity_store_structs.go#L114-L116
And this variable:
https://github.com/hashicorp/vault/blob/0d147185d6304838d4667ef7cb30d8ed3a41da3f/vault/identity_store.go#L34

**Comment by maxb** [comment]
Indeed, isn't acting as an invalidation signal the whole reason for this key's existence?

Are you willing/able to share more reason about this change publicly? Once upon a time (years ago), with a previous employer, I witnessed a spontaneous change in case sensitivity from False to True in a production cluster, and we never did figure out how that was even possible. I'm kind of curious whether more information has come to light! :-)

**Comment by akshya96** [comment]
> If we are removing the invalidation for this key in storage, should we also look to remove the key from storage in general moving forward? The only place we update this value in storage is here:
> 
> https://github.com/hashicorp/vault/blob/0d147185d6304838d4667ef7cb30d8ed3a41da3f/vault/identity_store.go#L649-L656
> 
> 
> Which looks to also be (after this change) the last reference for this struct:
> https://github.com/hashicorp/vault/blob/0d147185d6304838d4667ef7cb30d8ed3a41da3f/vault/identity_store_structs.go#L114-L116
> 
> 
> And this variable:
> https://github.com/hashicorp/vault/blob/0d147185d6304838d4667ef7cb30d8ed3a41da3f/vault/identity_store.go#L34
I agree that since the invalidation of the key is removed, we can remove the key itself but I think we will still need this field in the identitystore because it uses it to create the schema.

**Comment by ncabatoff** [comment]
> Are you willing/able to share more reason about this change publicly? Once upon a time (years ago), with a previous employer, I witnessed a spontaneous change in case sensitivity from False to True in a production cluster, and we never did figure out how that was even possible. I'm kind of curious whether more information has come to light! :-)

Until https://github.com/hashicorp/vault/pull/20964 we allowed for duplicate groups to be created due to a bug.  Once duplicate groups exist, on load we modify the in-memory setting, then during the initialize call of the identitystore we'll modify the stored setting.

**Comment by raskchanky** [comment]
I read the Slack discussion and the Jira and removing invalidation makes sense given the key isn't supposed to change, but I'm not following the part about removing the bits from `initialize()` that write the key to storage. If it's not in storage, where does the default value come from?

**Review by raskchanky — APPROVED** [review]
Based on the background info I read, I _think_ this is ok?

File: vault/identity_store.go [hunk 1]
```diff
@@ -645,15 +644,7 @@ func (i *IdentityStore) initialize(ctx context.Context, req *logical.Initializat
...
-
-	entry, err := logical.StorageEntryJSON(caseSensitivityKey, &casesensitivity{
-		DisableLowerCasedNames: i.disableLowerCasedNames,
-	})
-	if err != nil {
-		return err
-	}
-
-	return i.view.Put(ctx, entry)
+	return nil
```
Comment (ncabatoff) [inline vault/identity_store.go:647]: Should we delete the storage entry if it exists?

Comment (akshya96) [reply vault/identity_store.go:647]: I think upon upgrade, if the storage entry exists we can remove it as it is of no use anymore.

File: vault/identity_store.go [hunk 2]
```diff
@@ -646,14 +646,22 @@ func (i *IdentityStore) initialize(ctx context.Context, req *logical.Initializat
 		return err
 	}
 
-	entry, err := logical.StorageEntryJSON(caseSensitivityKey, &casesensitivity{
-		DisableLowerCasedNames: i.disableLowerCasedNames,
-	})
+	// if the storage entry for caseSensitivityKey exists, remove it
+	storageEntry, err := i.view.Get(ctx, caseSensitivityKey)
```
Comment (ncabatoff) [inline vault/identity_store.go:650]: You don't need a Get, you can just Delete - it won't return an error if the thing is already missing.

Comment (ncabatoff) [reply vault/identity_store.go:650]: Although now that I think about it, let's keep this, and add a log line where we record the value it had at time of deletion.  Could be useful in debugging at some point.

Comment (akshya96) [reply vault/identity_store.go:650]: added debug line to get caseSensitivityKey value at the time of deletion

File: vault/identity_store.go [hunk 3]
```diff
@@ -646,14 +646,32 @@ func (i *IdentityStore) initialize(ctx context.Context, req *logical.Initializat
...
+	}
+
+	if storageEntry != nil {
+		var setting casesensitivity
+		err := storageEntry.DecodeJSON(&setting)
+		switch err {
+		case nil:
+			i.logger.Debug("removing storage entry for case sensitivity key", "value", setting.DisableLowerCasedNames)
+		default:
+			i.logger.Error("failed to decode case sensitivity key", "error", err)
```
Comment (ncabatoff) [inline vault/identity_store.go:663]:
```suggestion
			i.logger.Error("failed to decode case sensitivity key, removing its storage entry anyway", "error", err)
```

I don't think we need an extra debug log line for this.

File: vault/identity_store_test.go [hunk 4]
```diff
@@ -857,3 +857,59 @@ func TestIdentityStore_UpdateAliasMetadataPerAccessor(t *testing.T) {
...
+		t.Fatalf("bad: expected a non-nil entry for casesensitivity key")
+	}
+
+	// Seal and unseal to trigger identityStore initialize
+	if err = c.Seal(root); err != nil {
+		t.Fatal(err)
+	}
+
+	var unsealed bool
+	for i := 0; i < 3; i++ {
```
Comment (ncabatoff) [inline vault/identity_store_test.go:896]: Better to range over the unsealKey slice than hardcode the number 3.

**Review by ncabatoff — APPROVED** [review]

**Pull request merged**